marker_uilints/tests/ui/expr/sugar/while_loop.stderr

```diff
@@ -0,0 +1,96 @@
+warning: expr
+  --> $DIR/while_loop.rs:8:5
+   |
+8  | /     while let Some(_) = something {
+9  | |         something = None;
+10 | |     }
+   | |_____^
+   |
+   = note: SpanSource: File(
+               FileInfo {
+                   file: "$DIR/while_loop.rs",
+                   span_src: SpanSrcId(..),
+               },
+           )
+   = note: Snippet: "while let Some(_) = something {/n        something = None;/n    }"
+note: the lint level is defined here
+  --> $DIR/while_loop.rs:7:12
+   |
+7  |     #[warn(marker::print_every_expr)]
+   |            ^^^^^^^^^^^^^^^^^^^^^^^^
+
+warning: expr
+ --> $DIR/while_loop.rs:8:11
+  |
+8 |     while let Some(_) = something {
```

Comment:
It's also surprising that the entire `let pat = scrutinee` is considered to be an expression.

Reply:
The reference doesn't consider it a separate expression, but this causes a lot of headache. For example `while` and `while let` loops are officially considered different expressions, it's the same with `if` expressions then there is also the question how [if-let-chains](https://rust-lang.github.io/rfcs/2497-if-let-chains.html) should be modeled without that being its own expression.

In practice rustc and I believe RA both model it as an expression and Marker does as well. I think the main reason for now having it an official expression, is that you could otherwise write weird code like
```
for x in {let x = 17; [1, 2]} {
}
```

Return expressions are in this weird group, you can write them basically everywhere, like in this example:

```
    if return () {
        println!("Hey")
    }
```

Reply:
I see. I don't think that the let part of the `while/if let` syntax being an expression is a problem at least today. I was interested in the reason why that is.

Regarding the return expression I think it is not surprising that you can write them everywhere. The `return` is simply an operator just like an unary negation (`!`). Everywhere where you can write an unary negation (`!`) you should be able to write a return. It is not a special case in this regard.